### PR TITLE
Allow ops to take primitive arrays, not just Tensors

### PR DESCRIPTION
**Pull request by dsmilkov**
FEATURE

This change allows users to pass primitives to operations, e.g. calling `tf.square(3)` instead of `tf.square(tf.scalar(3))`.

Also add deep shape checking when inferring shape of arrays of numbers. E.g.
`[[1, 2], [3], [5, 6]]` will fail to be parsed as Tensor because the 2nd (middle) element has only 1 value instead of 2.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1119)

### CODE REVIEW

**Comment by dsmilkov** [comment]
Reviewed 31 of 64 files at r1.
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[src/tensor_util_test.ts, line 131 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1119#-LFyYfvbDS3EtzEtKKHn:-LG0dtyx2_G3fj3bC-YE:bb6un1j) ([raw file](https://github.com/tensorflow/tfjs-core/blob/aa230aed3e9d7166002852db09a0f8b77bde4b25/src/tensor_util_test.ts#L131)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

would be good to also assert shape here and below
</blockquote></details>

Done

---

*[src/util.ts, line 117 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1119#-LFzCweM1yrtqXR6IZAk:-LG0e07G-VxA2VoWoxgU:bb6un1j) ([raw file](https://github.com/tensorflow/tfjs-core/blob/aa230aed3e9d7166002852db09a0f8b77bde4b25/src/util.ts#L117)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

how about "deepAssertShapeConsistency"
</blockquote></details>

Done

---

*[src/ops/array_ops.ts, line 359 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1119#-LFzDe2CELlc-r2L4kDV:-LG0gNX-1Xi7PFjBsnmF:bb6un1j) ([raw file](https://github.com/tensorflow/tfjs-core/blob/aa230aed3e9d7166002852db09a0f8b77bde4b25/src/ops/array_ops.ts#L359)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

any reason to throw here and not implement TensorLike? could actually be quite useful
</blockquote></details>

Done

---

*[src/ops/batchnorm_test.ts, line 320 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1119#-LFzERTQ9uzs1w8QD7Xs:-LG0gMET3vMQkOQ3ZuOL:bb6un1j) ([raw file](https://github.com/tensorflow/tfjs-core/blob/aa230aed3e9d7166002852db09a0f8b77bde4b25/src/ops/batchnorm_test.ts#L320)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

un-tensor these
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1119)*
